ls1intum/Artemis
# Improved plagiarism detection for text exercises

### Checklist

### Motivation and Context
Currently, the plagiarism split view to inspect similar student submissions is only available for modeling exercises. This PR allows instructors to also compare detected text submissions in the new plagiarism inspection UI.

### Description
This PR contains the following changes:
* Enhancements to the plagiarism inspection UI
    * New page header
    * Updated design of the Sidebar
    * Code improvements
* Display similar **text** submissions in the split view
* Use the new JPlag 3 API to detect plagiarism
* New plagiarism related classes on the client and server

The most notable change will be the updated plagiarism inspection UI and the possibility to compare similar text exercises. Additionally, some server code has been adapted to work with the new JPlag version and to use the added plagiarism types.

### Steps for Testing

To properly test this PR, you need a text exercise with at least two *similar* submissions.

Tip: Use the [Lorem Ipsum Generator](https://loremipsum.io/generator/) and slightly modify the generated text for each submission.

1. Log in to Artemis
2. Navigate to Course Administration
3. Select a course
4. Go to Exercises
5. Navigate to the details page of a text exercise by clicking 'View'
6. Click 'Check Plagiarism'

You should now see the plagiarism page. Make sure the following works:
* The sidebar items correctly display the student logins and similarity
* By selecting a sidebar item, the submissions in the split view update correctly
* You can confirm or deny a detected plagiarism
* You can download the plagiarism results both as JSON and CSV
* The downloaded plagiarism results contain the correct status ('confirmed', 'denied', 'none')
* You can resize the split view panes, both by dragging the gutter and clicking on the shortcut icons in the top right

Lastly, it would be great if you could verify that the plagiarism inspection for **programming** exercises still works. Programming exercises currently don't use the new plagiarism UI, but clicking on 'Check Plagiarism' should correctly download the results as a ZIP file.

**Note**: As mentioned in the reviews below, JPlag sometimes gives identical submissions a similarity slightly below 100%. I double-checked this and found out that this behavior already exists in JPlag 2 (previous version) and since I didn't make any changes to the algorithm itself, fixing this is beyond the scope of this PR.

### Test Coverage
plagiarism-header.component.ts                      100.00 %
plagiarism-inspector.component.ts                    90.91 %
plagiarism-sidebar.component.ts                       87.50 %
plagiarism-details.component.ts                        81.82 %
text-submission-viewer.component.ts              82.35 %
modeling-submission-viewer.component.ts     88.89 %
plagiarism-split-view.component.ts                  84.85 %

### Screenshots
![Screenshot 2020-11-30 at 22 10 13](https://user-images.githubusercontent.com/21973908/100754454-692fbf00-33eb-11eb-98a5-549effb2d35b.png)
![Screenshot 2020-11-30 at 22 10 39](https://user-images.githubusercontent.com/21973908/100754491-72b92700-33eb-11eb-978f-47f2b0fef96e.png)
